exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,262 @@
+)
+
+const (
+	prefixSplitter = "+"
+)
+
+type metricID *export.Descriptor
+
+// Exporter is an implementation of metric.Exporter that sends metrics to
+// Prometheus.
+type Exporter struct {
+	sync.RWMutex
+	registerer prometheus.Registerer
+	gatherer   prometheus.Gatherer
+
+	counters map[metricID]prometheus.Counter
+	gauges   map[metricID]prometheus.Gauge
+
+	onRegisterError func(e error)
+	counterVecs     map[string]*prometheus.CounterVec
+	gaugeVecs       map[string]*prometheus.GaugeVec
+}
+
+var _ export.Exporter = (*Exporter)(nil)
+
+// Options is a set of options for the tally reporter.
+type Options struct {
+	// Registerer is the prometheus registerer to register
+	// metrics with. Use nil to specify the default registerer.
+	Registerer prometheus.Registerer
+
+	// Gatherer is the prometheus gatherer to gather
+	// metrics with. Use nil to specify the default gatherer.
+	Gatherer prometheus.Gatherer
+
+	// DefaultHistogramBuckets is the default histogram buckets
+	// to use. Use nil to specify the default histogram buckets.
+	DefaultHistogramBuckets []float64
+
+	// DefaultSummaryObjectives is the default summary objectives
+	// to use. Use nil to specify the default summary objectives.
+	DefaultSummaryObjectives map[float64]float64
+
+	// OnRegisterError defines a method to call to when registering
+	// a metric with the registerer fails. Use nil to specify
+	// to panic by default when registering fails.
+	OnRegisterError func(err error)
+}
+
+// NewExporter returns a new prometheus exporter for prometheus metrics.
+func NewExporter(opts Options) (*Exporter, error) {
+	if opts.Registerer == nil {
+		opts.Registerer = prometheus.DefaultRegisterer
+	} else {
+		// A specific registerer was set, check if it's a registry and if
+		// no gatherer was set, then use that as the gatherer
+		if reg, ok := opts.Registerer.(*prometheus.Registry); ok && opts.Gatherer == nil {
+			opts.Gatherer = reg
+		}
+	}
+	if opts.Gatherer == nil {
+		opts.Gatherer = prometheus.DefaultGatherer
+	}
+	if opts.OnRegisterError == nil {
+		opts.OnRegisterError = func(err error) {
+			panic(err)
+		}
+	}
+
+	// TODO: should we make a "PullController" ?
+	go func() {
+		_ = http.ListenAndServe(":22022", promhttp.HandlerFor(opts.Gatherer, promhttp.HandlerOpts{}))
+	}()
+
+	return &Exporter{
+		registerer:      opts.Registerer,
+		gatherer:        opts.Gatherer,
+		onRegisterError: opts.OnRegisterError,
+
+		counters: make(map[metricID]prometheus.Counter),
+		gauges:   make(map[metricID]prometheus.Gauge),
+
+		counterVecs: make(map[string]*prometheus.CounterVec),
+		gaugeVecs:   make(map[string]*prometheus.GaugeVec),
+	}, nil
+}
+
+// Export exports the provide metric record to prometheus.
+func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+	var aggError error
+	checkpointSet.ForEach(func(record export.Record) {
+		agg := record.Aggregator()
+		if sum, ok := agg.(aggregator.Sum); ok {
```

Comment:
N.B. When we get to adding support for measure metric instruments, we'll want to put that code first, since the measure aggregator interfaces also implement `aggregator.Sum`.